src/main/java/edu/uwaterloo/ece658/service/DataHandlerService.java

```diff
@@ -0,0 +1,175 @@
+    
+    @EJB
+    private TagFacade tagFacade;
+
+    @EJB
+    private S3Service s3Service;
+    
+    private Photo createNewPhoto(
+            String s3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate, 
+            boolean isPrivatePhoto)
+    {
+        Photo photo = new Photo();        
+        photo.setS3Key(s3key);
+        photo.setMd5(Md5);
+        photo.setSize(photoSize);
+        photo.setUploadedTime(uploadDate);
+        photo.setIsPrivate(isPrivatePhoto);
+        photoFacade.create(photo);
+        return photo;
+    }
+    
+    private Tag createNewTag(String tagName) {
+        Tag tag = new Tag();
+        tag.setName(tagName);
+        tagFacade.create(tag);
+        return tag;
+    }
+    
+    private Tag initializeTag(String tagName) {
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        if (tag == null) {
+            tag = createNewTag(tagName);
+        }
+        return tag;
+    }
+    
+    private List<Tag> initializeTags(List<String> tagNames) {
+        List<Tag> tags = new ArrayList<>();
+        
+        for (String tagName : tagNames) {
+            tags.add(initializeTag(tagName));
+        }
+        return tags;
+    }
+    
+    public boolean updateDatabaseWithPhotoInformation(
+            String username, 
+            String s3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate,
+            boolean isPrivatePhoto,
+            List<String> tagNames)
+    {
+        List<Tag> tags = initializeTags(tagNames);
+        User user = userFacade.findUserByUserName(username);
+        Photo photo = createNewPhoto(
+                        s3key,
+                        Md5,
+                        photoSize,
+                        uploadDate,
+                        isPrivatePhoto);
+        
+        photo.setTags(tags);
+        photo.addUploadedUser(user);
+        photoFacade.edit(photo);
+        
+        user.addUploadedPhoto(photo);
+        userFacade.edit(user);
+        
+        for (Tag tag : tags) {
+            tag.addPhotoUnderThisTag(photo);
+            tagFacade.edit(tag);
+        }
+        
+        // call notify to send messages to users
+        if (!isPrivatePhoto) {
+            return subscriptionService.notifyUsersBySubscribedTags(tags, photo);
+        }
+        
+        return true;
+    }
+    
+    public void removeTagFromPhoto(Photo photo, Tag tag) {        
+        photo.removeTag(tag);
+        tag.removePhotoUnderThisTag(photo);
+        tagFacade.edit(tag);
+        photoFacade.edit(photo);
+    }
+    
+    public void updatePhotoTags(String username, String s3Key, List<String> newTags) {
+        Photo photo = photoFacade.getPhotoByS3Key(s3Key);
+        User user = userFacade.findUserByUserName(username);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            List<String> tagNames = new ArrayList<>();

```

Comment:
Why not simply assign `photo.getTags()` instead of creating a new list and add all existing tags?

Reply:
Oh, I understand now. Because you have removed some tags from the original list. Let's merge it!